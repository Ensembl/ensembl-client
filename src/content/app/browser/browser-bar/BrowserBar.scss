@import 'src/styles/common';

.browserBar {
  display: flex;
  width: 100%;
  align-items: center;
  position: relative;
}

.browserResetWrapper {
  margin-right: 14px;
}

.browserLocationIndicatorWrapper {
  margin-left: auto;
}
