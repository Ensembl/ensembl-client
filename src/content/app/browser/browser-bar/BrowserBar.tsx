/**
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { useSelector } from 'react-redux';

import { getChrLocation, getBrowserActiveEnsObject } from '../browserSelectors';
import { getIsDrawerOpened } from 'src/content/app/browser/state/drawer/drawerSelectors';

import BrowserReset from '../browser-reset/BrowserReset';
import FeatureSummaryStrip from 'src/shared/components/feature-summary-strip/FeatureSummaryStrip';
import BrowserLocationIndicator from '../browser-location-indicator/BrowserLocationIndicator';

import { ChrLocation } from '../browserState';
import { EnsObject } from 'src/shared/state/ens-object/ensObjectTypes';

import styles from './BrowserBar.scss';
import { Environment, isEnvironment } from 'src/shared/helpers/environment';

export type BrowserBarProps = {
  chrLocation: ChrLocation | null;
  defaultChrLocation: ChrLocation | null;
  isDrawerOpened: boolean;
  ensObject: EnsObject | null;
};

export const BrowserBar = () => {
  const chrLocation = useSelector(getChrLocation);
  const ensObject = useSelector(getBrowserActiveEnsObject);
  const isDrawerOpened = useSelector(getIsDrawerOpened);

  // return empty div instead of null, so that the dedicated slot in the CSS grid of StandardAppLayout
  // always contains a child DOM element
  if (!(chrLocation && ensObject)) {
    return <div />;
  }

  return (
    <div className={styles.browserBar}>
      <div className={styles.browserResetWrapper}>
        <BrowserReset />
      </div>
      {ensObject && (
        <FeatureSummaryStrip ensObject={ensObject} isGhosted={isDrawerOpened} />
      )}
      <div className={styles.browserLocationIndicatorWrapper}>
        <BrowserLocationIndicator
          disabled={isDrawerOpened || isEnvironment([Environment.PRODUCTION])}
        />
      </div>
    </div>
  );
};

export default BrowserBar;
