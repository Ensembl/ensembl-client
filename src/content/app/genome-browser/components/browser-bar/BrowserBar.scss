@import 'src/styles/common';

.browserBar {
  display: flex;
  width: 100%;
  align-items: center;
  position: relative;
  padding-left: 18px;
}

.browserResetWrapper {
  margin-right: 20px;
}

.browserLocationIndicatorWrapper {
  margin-left: auto;
}
