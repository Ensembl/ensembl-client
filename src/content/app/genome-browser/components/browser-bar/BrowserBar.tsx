/**
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { useSelector } from 'react-redux';

import { Environment, isEnvironment } from 'src/shared/helpers/environment';

import BrowserReset from '../browser-reset/BrowserReset';
import FeatureSummaryStrip from 'src/shared/components/feature-summary-strip/FeatureSummaryStrip';
import BrowserLocationIndicator from '../browser-location-indicator/BrowserLocationIndicator';

import { getIsDrawerOpened } from 'src/content/app/genome-browser/state/drawer/drawerSelectors';
import { getBrowserActiveEnsObject } from 'src/content/app/genome-browser/state/browser-general/browserGeneralSelectors';
import { getChrLocation } from 'src/content/app/genome-browser/state/browser-location/browserLocationSelectors';

import { EnsObject } from 'src/shared/state/ens-object/ensObjectTypes';
import { ChrLocation } from 'src/content/app/genome-browser/state/browser-location/browserLocationSlice';

import styles from './BrowserBar.scss';

export type BrowserBarProps = {
  chrLocation: ChrLocation | null;
  defaultChrLocation: ChrLocation | null;
  isDrawerOpened: boolean;
  ensObject: EnsObject | null;
};

export const BrowserBar = () => {
  const chrLocation = useSelector(getChrLocation);
  const ensObject = useSelector(getBrowserActiveEnsObject);
  const isDrawerOpened = useSelector(getIsDrawerOpened);

  // return empty div instead of null, so that the dedicated slot in the CSS grid of StandardAppLayout
  // always contains a child DOM element
  if (!(chrLocation && ensObject)) {
    return <div />;
  }

  return (
    <div className={styles.browserBar}>
      <div className={styles.browserResetWrapper}>
        <BrowserReset />
      </div>
      {ensObject && (
        <FeatureSummaryStrip ensObject={ensObject} isGhosted={isDrawerOpened} />
      )}
      <div className={styles.browserLocationIndicatorWrapper}>
        <BrowserLocationIndicator
          disabled={isDrawerOpened || isEnvironment([Environment.PRODUCTION])}
        />
      </div>
    </div>
  );
};

export default BrowserBar;
