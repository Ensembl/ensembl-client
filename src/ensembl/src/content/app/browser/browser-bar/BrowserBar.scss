@import 'src/styles/common';

.browserBar {
  @include flex;
  background: $ens-light-grey;
  box-shadow: 0 2px 3px $ens-grey;
  height: 32px;
  padding: 7px 0 3px 0;
  position: relative;
  z-index: 200;

  dd {
    display: inline-block;
    margin: 0 18px;
  }

  button {
    &.selected {
      background: $ens-white;
    }
  }
}

.browserInfo {
  @include flex;
  flex-wrap: nowrap;
  font-size: 12px;
  margin: 0;
  width: calc(100vw - 36px);

  @include breakpoint(large) {
    width: calc(100vw - 356px);
  }

  label {
    color: $ens-dark-grey;
    font-weight: $light;
  }

  .value {
    color: $ens-black;
    margin: 0 3px;
  }

  .nonLabelValue {
    text-transform: lowercase;
  }

  &.browserInfoExpanded {
    width: calc(100vw - 36px);
  }

  &.browserInfoGreyed {
    .value,
    .nonLabelValue {
      color: $ens-dark-grey;
    }
  }

  dd {
    margin-right: 10px;
    white-space: nowrap;
  }

  dl {
    overflow: hidden;

    &.browserInfoLeft {
      flex: 1 1 auto;
    }

    &.browserInfoRight {
      flex: 0 1 auto;
      text-align: right;
    }

    .ensObjectLabel {
      margin-left: 14px;
      text-transform: capitalize;

      .value {
        font-size: 13px;
        font-weight: $bold;
      }
    }
  }

  .navigator {
    margin: 0 10px 0 0;
    position: relative;
    top: -2px;

    img {
      height: 22px;
      width: 22px;
    }
  }
}
