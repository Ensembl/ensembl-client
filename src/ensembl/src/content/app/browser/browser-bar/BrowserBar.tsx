import React, { FunctionComponent, useState, useEffect } from 'react';
import { connect } from 'react-redux';
import classNames from 'classnames';

import { browserInfoConfig, BrowserInfoItem } from '../browserConfig';
import { TrackType } from '../track-panel/trackPanelConfig';

import { toggleBrowserNav, toggleGenomeSelector } from '../browserActions';
import { ChrLocation } from '../browserState';
import {
  getBrowserNavOpened,
  getChrLocation,
  getBrowserActivated,
  getDefaultChrLocation,
  getGenomeSelectorActive,
  getBrowserActiveGenomeId
} from '../browserSelectors';
import { getDrawerOpened } from '../drawer/drawerSelectors';
import { getEnsObjectInfo } from 'src/ens-object/ensObjectSelectors';
import {
  getSelectedBrowserTab,
  getTrackPanelModalOpened,
  getTrackPanelOpened
} from '../track-panel/trackPanelSelectors';
import { selectBrowserTabAndSave } from '../track-panel/trackPanelActions';
import { toggleDrawer } from '../drawer/drawerActions';
import { RootState } from 'src/store';
import { EnsObject } from 'src/ens-object/ensObjectTypes';

import BrowserReset from '../browser-reset/BrowserReset';
import BrowserGenomeSelector from '../browser-genome-selector/BrowserGenomeSelector';
import BrowserTabs from '../browser-tabs/BrowserTabs';

import styles from './BrowserBar.scss';

type StateProps = {
  activeGenomeId: string;
  browserActivated: boolean;
  browserNavOpened: boolean;
  chrLocation: { [genomeId: string]: ChrLocation };
  defaultChrLocation: { [genomeId: string]: ChrLocation };
  drawerOpened: boolean;
  genomeSelectorActive: boolean;
  ensObjectInfo: EnsObject;
  selectedBrowserTab: { [genomeId: string]: TrackType };
  trackPanelModalOpened: boolean;
  trackPanelOpened: boolean;
};

type DispatchProps = {
  selectBrowserTabAndSave: (selectedBrowserTab: TrackType) => void;
  toggleBrowserNav: () => void;
  toggleDrawer: (drawerOpened: boolean) => void;
  toggleGenomeSelector: (genomeSelectorActive: boolean) => void;
};

type OwnProps = {
  dispatchBrowserLocation: (chrLocation: ChrLocation) => void;
};

type BrowserBarProps = StateProps & DispatchProps & OwnProps;

type BrowserInfoProps = {
  ensObjectInfo: EnsObject;
};

type BrowserNavigatorButtonProps = {
  toggleNavigator: () => void;
  navigator: BrowserInfoItem;
  icon: string; // TODO: use inline SVG
};

export const BrowserBar: FunctionComponent<BrowserBarProps> = (
  props: BrowserBarProps
) => {
  const shouldShowBrowserInfo = () => {
    const chrLocationForGenome = props.defaultChrLocation[props.activeGenomeId];
    const isLocationOfWholeChromosome =
      chrLocationForGenome === undefined ? true : false;

    return !(props.genomeSelectorActive || isLocationOfWholeChromosome);
  };

  const { navigator } = browserInfoConfig;
  const [showBrowserInfo, toggleShowBrowserInfo] = useState(
    shouldShowBrowserInfo()
  );

  const setBrowserInfoVisibility = () => {
    const shouldToggleVisibility = showBrowserInfo !== shouldShowBrowserInfo();
    if (shouldToggleVisibility) {
      toggleShowBrowserInfo(!showBrowserInfo);
    }
  };

  useEffect(() => {
    setBrowserInfoVisibility();
  }, [props.defaultChrLocation, props.genomeSelectorActive]);

  const getBrowserNavIcon = () => {
    if (props.drawerOpened === true) {
      return navigator.icon.grey as string;
    } else if (props.browserNavOpened === true) {
      return navigator.icon.selected as string;
    } else {
      return navigator.icon.default;
    }
  };

  const toggleNavigator = () => {
    if (props.drawerOpened === true) {
      return;
    }

    props.toggleBrowserNav();
  };

  const className = classNames(styles.browserInfo, {
    [styles.browserInfoExpanded]: !props.trackPanelOpened,
    [styles.browserInfoGreyed]: props.drawerOpened
  });

  return (
    <div className={styles.browserBar}>
      <div className={className}>
        <dl className={styles.browserInfoLeft}>
          <BrowserReset
            activeGenomeId={props.activeGenomeId}
            dispatchBrowserLocation={props.dispatchBrowserLocation}
            chrLocation={props.chrLocation}
            defaultChrLocation={props.defaultChrLocation}
            drawerOpened={props.drawerOpened}
          />
          {showBrowserInfo && (
            <BrowserInfo ensObjectInfo={props.ensObjectInfo} />
          )}
        </dl>
        <dl className={styles.browserInfoRight}>
          <BrowserGenomeSelector
            activeGenomeId={props.activeGenomeId}
            browserActivated={props.browserActivated}
            dispatchBrowserLocation={props.dispatchBrowserLocation}
            chrLocation={props.chrLocation}
            drawerOpened={props.drawerOpened}
            genomeSelectorActive={props.genomeSelectorActive}
            toggleGenomeSelector={props.toggleGenomeSelector}
          />
          {!props.genomeSelectorActive && props.ensObjectInfo.genome_id && (
            <BrowserNavigatorButton
              navigator={navigator}
              toggleNavigator={toggleNavigator}
              icon={getBrowserNavIcon()}
            />
          )}
        </dl>
      </div>
      {props.trackPanelOpened && (
        <BrowserTabs
          activeGenomeId={props.activeGenomeId}
          ensObjectInfo={props.ensObjectInfo}
          drawerOpened={props.drawerOpened}
          genomeSelectorActive={props.genomeSelectorActive}
          selectBrowserTabAndSave={props.selectBrowserTabAndSave}
          selectedBrowserTab={props.selectedBrowserTab}
          toggleDrawer={props.toggleDrawer}
          trackPanelModalOpened={props.trackPanelModalOpened}
        />
      )}
    </div>
  );
};

export const BrowserInfo = ({ ensObjectInfo }: BrowserInfoProps) => (
  <>
    {ensObjectInfo.object_type === 'gene' && (
      <>
        <dd className={styles.ensObjectLabel}>
          <label>{ensObjectInfo.object_type}</label>
          <span className={styles.value}>{ensObjectInfo.label}</span>
        </dd>
        <dd>
          <label>Stable ID</label>
          <span className={styles.value}>{ensObjectInfo.stable_id}</span>
        </dd>
        <dd className="show-for-large">
          <label>Spliced mRNA length</label>
          <span className={styles.value}>{ensObjectInfo.spliced_length}</span>
          <label>bp</label>
        </dd>
        <dd className={`show-for-large ${styles.nonLabelValue}`}>
          {ensObjectInfo.bio_type}
        </dd>
        <dd className={`show-for-large ${styles.nonLabelValue}`}>
          {ensObjectInfo.strand} strand
        </dd>
      </>
    )}

    {ensObjectInfo.object_type === 'region' && (
      <>
        <dd className={styles.ensObjectLabel}>
          <label>Region: </label>
          <span
            className={styles.value}
          >{`${ensObjectInfo.location.chromosome}:${ensObjectInfo.location.start}:${ensObjectInfo.location.end}`}</span>
        </dd>
      </>
    )}
  </>
);

export const BrowserNavigatorButton = (props: BrowserNavigatorButtonProps) => (
  <dd className={styles.navigator}>
    <button title={props.navigator.description} onClick={props.toggleNavigator}>
      <img src={props.icon} alt={props.navigator.description} />
    </button>
  </dd>
);

const mapStateToProps = (state: RootState): StateProps => ({
  activeGenomeId: getBrowserActiveGenomeId(state),
  browserActivated: getBrowserActivated(state),
  browserNavOpened: getBrowserNavOpened(state),
  chrLocation: getChrLocation(state),
  defaultChrLocation: getDefaultChrLocation(state),
  drawerOpened: getDrawerOpened(state),
  ensObjectInfo: getEnsObjectInfo(state),
  genomeSelectorActive: getGenomeSelectorActive(state),
  selectedBrowserTab: getSelectedBrowserTab(state),
  trackPanelModalOpened: getTrackPanelModalOpened(state),
  trackPanelOpened: getTrackPanelOpened(state)
});

const mapDispatchToProps: DispatchProps = {
  selectBrowserTabAndSave,
  toggleBrowserNav,
  toggleDrawer,
  toggleGenomeSelector
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(BrowserBar);
