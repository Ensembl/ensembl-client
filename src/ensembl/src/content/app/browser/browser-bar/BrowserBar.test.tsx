import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import BrowserBar from './BrowserBar';

describe('<BrowserBar />', () => {
  let toggleBrowserNavFn: () => void;
  let wrapper: any;

  beforeEach(() => {
    toggleBrowserNavFn = jest.fn();
    wrapper = shallow(
      <BrowserBar
        browserNavOpened={true}
        expanded={true}
        toggleBrowserNav={toggleBrowserNavFn}
      />
    );
  });

  test('has a left bar', () => {
    expect(wrapper.find('.browserInfoLeft')).toHaveLength(1);
  });

  test('has a right bar', () => {
    expect(wrapper.find('.browserInfoRight')).toHaveLength(1);
  });

  test('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
