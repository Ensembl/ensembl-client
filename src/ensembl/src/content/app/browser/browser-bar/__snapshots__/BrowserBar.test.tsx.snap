// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BrowserBar /> renders correctly 1`] = `undefined`;
